### ISSUE TITLE
StringTools.startsWith code for Python is suboptimal

### ISSUE THREAD

**Issue by techtonik** [HaxeFoundation/haxe#6138]
Compile this into Python with `haxe -main HelloWorld -python hello.py`:
```
class HelloWorld {
  static public function main() {
    trace(StringTools.startsWith('Hello World', 'Hello'));
  }
}
```
The result is big:
```
# Generated by Haxe 3.4.2 (git build master @ 890f8c7)
# coding: utf-8

class HelloWorld:
    __slots__ = ()

    @staticmethod
    def main():
        print(str(StringTools.startsWith("Hello World","Hello")))

class StringTools:
    __slots__ = ()

    @staticmethod
    def startsWith(s,start):
        if (len(s) >= len(start)):
            return (HxString.substr(s,0,len(start)) == start)
        else:
            return False

class HxString:
    __slots__ = ()

    @staticmethod
    def substr(s,startIndex,_hx_len = None):
        if (_hx_len is None):
            return s[startIndex:]
        else:
            if (_hx_len == 0):
                return ""
            return s[startIndex:(startIndex + _hx_len)]

HelloWorld.main()
```
Why ideally it should be just:
```
# Generated by Haxe 3.4.2 (git build master @ 890f8c7)
# coding: utf-8

class HelloWorld:
    __slots__ = ()

    @staticmethod
    def main():
        print(str("Hello World".startswith("Hello")))

HelloWorld.main()
```

**Comment by techtonik** [comment]
Nice. I was trying to make the change myself, but couldn't understand the connection between .hx and .py files. Is there a tutorial that explains this change?

**Comment by nadako** [comment]
>  the connection between .hx and .py files.

Not sure I understood the question. Haxe generates a .py file, what I did is make it generate a simple field access on strings for startsWith/endsWith, so it calls native methods instead of providing its own implementation.

**Comment by techtonik** [comment]
The question is about this piece of code from `std/python/NativeStringTools.hx` which is called as `python.NativeStringTools.startswith(s, start);`:
````
	public static inline function startswith (s:String, prefix:String):Bool
	{
		return python.Syntax.field(s, "startswith")(prefix);
	}
````

Why it is `inline`? Where is the `python.Syntax.field` and other mapping conventions are described? Is there a tutorial or cookbook with more usage examples?

**Comment by nadako** [comment]
> Why it is inline?

Well that's function [inlining](http://haxe.org/manual/class-field-inline.html), meaning that calls to this function will be inlined (so instead of `NativeStringTools.startswith(a, b)` we'll have `python.Syntax.field(a, "startswith")(b)` expression generated).

> Where is the python.Syntax.field and other mapping conventions are described?

I don't think they are. `python.Syntax` could indeed use some documentation (and cleanup as well). I'll work on that when I carve some time.

Basically, it's a "magic" class, handled by the python output generator to generate python syntax directly from haxe. Knowing that it's fairly easy to deduce that `python.Syntax.field` generates a field access python syntax, so what this functions does is basically generates `s.startswith(prefix)`, where `s` and `prefix` are inlined arguments, so a call to`NativeStringTools.startswith(a, b)` will result in `a.startswith(b)` being generated. Hope that made it more clear!

**Comment by techtonik** [comment]
Yes. Now it is clear. Especially after trying to solve problem myself and then looking at solution. It could make my day though if I could discover a ticket like this yesterday. =) It is good as tutorial already, given that more details about `python.Syntax` magic are explained for those how need to know more.
